src/xrpld/ledger/detail/View.cpp

```diff
@@ -2123,4 +2492,65 @@ rippleCredit(
         saAmount.asset().value());
 }
 
+static Number
+getShareTotal(ReadView const& view, std::shared_ptr<SLE> const& vault)
+{
+    auto issuance =
+        view.read(keylet::mptIssuance(vault->at(sfMPTokenIssuanceID)));
+    return issuance->at(sfOutstandingAmount);
+}
+
+[[nodiscard]] STAmount
+assetsToSharesDeposit(
+    ReadView const& view,
+    std::shared_ptr<SLE> const& vault,
+    STAmount const& assets)
+{
+    assert(assets.asset() == vault->at(sfAsset));
+    Number assetTotal = *vault->at(sfAssetTotal);
+    STAmount shares{
+        vault->at(sfMPTokenIssuanceID), static_cast<Number>(assets)};
+    if (assetTotal == 0)
+        return shares;
+    Number shareTotal = getShareTotal(view, vault);
+    shares = shareTotal * (assets / assetTotal);
+    return shares;
+}
+
+[[nodiscard]] STAmount
+assetsToSharesWithdraw(
+    ReadView const& view,
+    std::shared_ptr<SLE> const& vault,
+    STAmount const& assets)
+{
+    assert(assets.asset() == vault->at(sfAsset));
+    Number assetTotal = vault->at(sfAssetTotal);
+    assetTotal -= vault->at(sfLossUnrealized);
+    STAmount shares{vault->at(sfMPTokenIssuanceID)};
+    if (assetTotal == 0)
+        return shares;
+    Number shareTotal = getShareTotal(view, vault);
+    shares = shareTotal * (assets / assetTotal);
+    // TODO: Limit by withdrawal policy?
+    return shares;
+}
+
+[[nodiscard]] STAmount
+sharesToAssetsWithdraw(
+    ReadView const& view,
+    std::shared_ptr<SLE> const& vault,
+    STAmount const& shares)
+{
+    assert(shares.asset() == vault->at(sfMPTokenIssuanceID));
+    Number assetTotal = vault->at(sfAssetTotal);
+    assetTotal -= vault->at(sfLossUnrealized);
+    STAmount assets{vault->at(sfAsset)};
+    if (assetTotal == 0)
+        return assets;
+    Number shareTotal = getShareTotal(view, vault);
+    assets = assetTotal * (shares / shareTotal);
+    // TODO: Limit by withdrawal policy?
```

Comment:
This has to be addressed.

Reply:
Nothing to be done, we only support FIFO now. Will remove this comment.